[documentation] Give more details on coloring algorithms (#297)
JuliaSmoothOptimizers/ADNLPModels.jl: docs/src/sparse.md

Prior to version 0.8.0, the default detector was `SymbolicSparsityDetector()` from `Symbolics.jl`.

- A **`coloring_algorithm`** must be of type `SparseMatrixColorings.GreedyColoringAlgorithm`.
The default algorithm is `GreedyColoringAlgorithm{:direct}()` for `SparseADJacobian` and `SparseADHessian`, while it is `GreedyColoringAlgorithm{:substitution}()` for `SparseReverseADHessian`.
These algorithms are available in the package `SparseMatrixColorings.jl`.

The `GreedyColoringAlgorithm{:direct}()` performs column coloring for Jacobians and star coloring for Hessians.
In contrast, `GreedyColoringAlgorithm{:substitution}()` applies acyclic coloring for Hessians.

> Substitution coloring results in fewer colors and thus requires fewer directional derivatives to recover all non-zeros of the sparse Hessian.

Reviewer: ```suggestion
The `:substitution` coloring mode usually finds fewer colors than the `:direct` mode and thus fewer directional derivatives are needed to recover all non-zeros of the sparse Hessian.
```

> However, it necessitates storing the entire compressed sparse Hessian.
> On the other hand, direct coloring only requires storing one column of the compressed Hessian at a time.

Reviewer: ```suggestion
However, it requires storing the compressed sparse Hessian, while `:direct` coloring only stores one column of the compressed Hessian.
```

> If the Hessian is highly ill-conditioned, direct coloring can be preferable because it avoids the need for triangular solves to recover the non-zeros from the compressed Hessian.

Reviewer: ```suggestion
The `:direct` coloring mode is numerically more stable and may be preferable for highly ill-conditioned Hessian as it doesn't require solving triangular systems to compute the non-zeros from the compressed Hessian.
```
